We first create the gtf file with the highest expressed two isoforms and TPM >=0.5. We next filtered out those with different CDS start (i.e. if two isoforms have different CDS start, only the one with the earlier star will be used). The reason to do this is because the main ORFs from the early start isoforms could be wrongly identified as uORFs in some cases.  

The gtf and other files required for this analysis is available at https://data.mendeley.com/preview/89j7snbm2r  

You will need to change the path accordingly.  
```
#############################################
#############################################
#Section 1: Find translated minimum uORFs   #
# Map Ribo-seq reads to each Minimum uORFs  #
#############################################
#############################################
rm(list=ls())
library(dplyr)
library(GenomicFeatures)
library(GenomicRanges)
library(Biostrings)
library(Rsamtools)
library(ORFik)
library(RiboPlotR)
library(future.apply)
library(ggplot2)
library(clipr)
library(openxlsx)
GTF_path <- "~/Desktop/CTRL_TPC/data/Araport11_20181206_max2_expressed_isoform_TPM0.5_rm_diff_CDSstart.gtf"
FA <- FaFile("~/Desktop/Leaky_scanning/TAIR10_chr_all_2.fas")
txdb <- makeTxDbFromGFF(file = GTF_path,format="gtf", dataSource="Araport11",organism="Arabidopsis")
exonByTx <- exonsBy(txdb, by="tx", use.names=T)
length(exonByTx) #21845
fiveUTR <- fiveUTRsByTranscript(txdb, use.names=T)
fiveUTR_seqs <- extractTranscriptSeqs(FA,fiveUTR)
fiveUTR_ORFs <- findMapORFs(fiveUTR, fiveUTR_seqs,startCodon = "ATG",longestORF=T,groupByTx=F)
fiveUTR_gene_id <- substr(names(fiveUTR),1,9)
fiveUTR_ORFs_gene_id <- substr(names(fiveUTR_ORFs),1,9)
fiveUTR_no_ORFs_gene_id <-  setdiff(fiveUTR_gene_id,fiveUTR_ORFs_gene_id)

length(fiveUTR) #20938
length(fiveUTR_ORFs) #28228
uORF_length <- sum(width(fiveUTR_ORFs))
fiveUTR_ORFs_seq <- extractTranscriptSeqs(FA,fiveUTR_ORFs)

#Make Ribo-seq reads to GRanges
Ribo1=read.delim(file="~/Desktop/CTRL_v1/CTRL_expressed_P_sites_sort_count",header=F,stringsAsFactors=F,sep="\t")
colnames(Ribo1) <- c("count","chr","start","strand")
RiboN <- Ribo1
RiboN$end <- RiboN$start
head(RiboN)
RiboR <- makeGRangesFromDataFrame(RiboN,keep.extra.columns=T)
RiboR

#XAA1 is a function to map Ribo-seq reads to each nt positions of minimum uORFs
#XAA1 is for minimum uORF and XAA (see below) is for >=2aa uORFs 
#This function also consider the blue reads for the -1 codon before start as in-frame reads
XAA1 <-function(x,y){
  # print(1)
  STRAND=as.character(strand(fiveUTR_ORFs_xaa[[x]])[1])
  if(STRAND=="+") {z=fiveUTR_ORFs_xaa[[x]]} 
  else if (STRAND=="-"){z=rev(fiveUTR_ORFs_xaa[[x]])}
  aa=unlist(Map(`:`, start(z), end(z)))
  gr=GRanges(seqnames=suppressWarnings(as.character(seqnames(z))[1]), ranges=IRanges(start = aa,width=1),strand=STRAND)
  # print(2)
  if(STRAND=="+"){
    gr$frame=rep(1:3,time=length(aa)/3)
    gr1 <- gr[1:3]
    gr2 <- gr[4:6]
    
  } else if (STRAND=="-"){
    gr$frame=rep(3:1,time=length(aa)/3)
    gr1 <- gr[6:4]
    gr2 <- gr[3:1]
  }
  # print(3)
  df0 <- data.frame(count=c(0,0,0),frame=c(1,2,3))
  ####***
  ranges1 <- subsetByOverlaps(RiboR,gr1)
  hits1 <- findOverlaps(RiboR, gr1)
  xframe1 <- unlist(IntegerList(split(gr1$frame[subjectHits(hits1)], queryHits(hits1))))
  mcols(ranges1) <- DataFrame(mcols(ranges1), xframe1)
  # print(4)
  df1 <- data.frame(count=ranges1$count,frame=ranges1$xframe1)
  # print(5)
  df1 <- rbind(df1,df0)
  # print(6)
  df1t1 <- table(df1$frame) %>% as.data.frame()
  # print(7)
  df1t2 <- df1 %>% group_by(frame) %>% 
    summarise(total_counts = sum(count)) %>% as.data.frame()
  # print(8)
  df1t2$frame_freq <- df1t1$Freq-1
  # print(9)
  ####***
  ranges2 <- subsetByOverlaps(RiboR,gr2)
  # print(10)
  hits2 <- findOverlaps(RiboR, gr2)
  xframe2 <- unlist(IntegerList(split(gr2$frame[subjectHits(hits2)], queryHits(hits2))))
  mcols(ranges2) <- DataFrame(mcols(ranges2), xframe2)
  df2 <- data.frame(count=ranges2$count,frame=ranges2$xframe2)
  df2 <- rbind(df2,df0)
  df2t1 <- table(df2$frame) %>% as.data.frame()
  df2t2 <- df2 %>% group_by(frame) %>% 
    summarise(total_counts = sum(count)) %>% as.data.frame()
  df2t2$frame_freq <- df2t1$Freq-1
  # print(16)
  ####***
  DF <- cbind(df1t2,df2t2[,2:3])
  colnames(DF) <- c("Frame","Counts1","FrameFreq1","CountStop","FrameFreqStop")
  DF
}

fiveUTR_ORFs_xaa <- fiveUTR_ORFs[uORF_length==(1+1)*3]
length(fiveUTR_ORFs_xaa) #1876

plan(multisession)
#future_lapply use multiple core for analysis
B=future_lapply(seq_len(length(fiveUTR_ORFs_xaa)),function(x) XAA1(x,y=1))
names(B) <- names(fiveUTR_ORFs_xaa)
#dir.create("~/Desktop/CTRL_TPC/data/Rdata/Mar02_2023/")
save(B,file=paste0("~/Desktop/CTRL_TPC/data/Rdata/Mar02_2023/2Max_isoform_0.5TPM_AA_",1,"_.RData"))
load(paste0("~/Desktop/CTRL_TPC/data/Rdata/Mar02_2023/2Max_isoform_0.5TPM_AA_",1,"_.RData"))

########
### Functions to get counts from different frames or total counts
InFrameCounts <- function(x,A){
  N=A[x]
  ToTal_in_frame_counts_wo_stop <- sum(N[[1]][c(1,2),2])
}

TotalCounts <- function(x,A){
  N=A[x]
  ToTal_counts_wo_stop <-sum(N[[1]][,2])
}

### Functions to get in frames sites with Ribo-seq reads or total sites
InFrameSites <- function(x,A){
  N=A[x]
  ToTal_in_frame_sites_wo_stop <- sum(N[[1]][c(1,2),3])
}

TotalSites <- function(x,A){
  N=A[x]
  ToTal_sites_wo_stop <-sum(N[[1]][,3])
}

#Function to extract Minimum uORFs with >=10 in frame reads and in frame percentage >50  
InFrameCountSites <- function(z){
  print(z)
  load(paste0("~/Desktop/CTRL_TPC/data/Rdata/2Max_isoform_0.5TPM_AA_",z,"_.RData"))
  gene_id <- substr(names(B),1,9)
  InFrameC_Mu <<- unlist(lapply(seq_len(length(B)), function(x) InFrameCounts(x,A=B)))
  TotalC_Mu <<- unlist(lapply(seq_len(length(B)), function(x) TotalCounts(x,A=B)))
  InFrameS_Mu <<- unlist(lapply(seq_len(length(B)), function(x) InFrameSites(x,A=B)))
  TotalS_Mu <<- unlist(lapply(seq_len(length(B)), function(x) TotalSites(x,A=B)))
  STRAND <<- sapply(seq_len(length(B)), function(x)as.character(strand(fiveUTR_ORFs_xaa[names(B)[x]])[[1]][1]))
  INFO <<- data.frame(ORF_id =names(B),
                      gene_id,
                      InFrameC=InFrameC_Mu,
                      TotalC=TotalC_Mu,
                      InFrameC_perc = round(InFrameC_Mu/TotalC_Mu*100,1),
                      InFrameS=InFrameS_Mu,
                      TotalS=TotalS_Mu,
                      InFrameS_perc = round(InFrameS_Mu/TotalS_Mu*100,1),
                      Strand = STRAND,
                      Mu=z)
  
  INFO[is.na(INFO)] <- 0
  INFO2 <<- INFO %>% filter(InFrameC_Mu>=10,InFrameC_perc>50) #TotalC>10,InFrameC_perc>50
  # Get 427
  # INFO2 <- INFO %>% filter(InFrameC_Mu>=10) 
  print(nrow(INFO2))
  INFO2
}

# Calculate the counts in minimum uORFs
Minimum_uORF <- InFrameCountSites(z=1)
Minimum_uORF$tx_id <- sub("_.*","",Minimum_uORF$ORF_id)
Minimum_uORF$chr <- substr(Minimum_uORF$ORF_id,3,3)

# Define coordinates for minimum uORF start or stop on transcripts
ORF_id <- Minimum_uORF$ORF_id
Tx_id <- gsub("\\_.*","",ORF_id)

uORF_Tx_START <- function(x){
  TxCord <- mapToTranscripts(fiveUTR_ORFs_xaa[[ORF_id[x]]], exonByTx[Tx_id[x]], ignore.strand = FALSE)
  start(TxCord[1])
}

uORF_Tx_END <- function(x){
  TxCord <- mapToTranscripts(fiveUTR_ORFs_xaa[[ORF_id[x]]], exonByTx[Tx_id[x]], ignore.strand = FALSE)
  end(TxCord[length(TxCord)])
}

#Find tx ranges for minimum uORFs
# plan(multisession)
uORF_Tx_START_data <- sapply(seq_len(length(ORF_id)),function(x) uORF_Tx_START(x))
uORF_Tx_END_data <- sapply(seq_len(length(ORF_id)),function(x) uORF_Tx_END(x))
Minimum_uORF$uORF_Tx_range <- paste0(uORF_Tx_START_data,"-",uORF_Tx_END_data)

#Find genomic ranges for minimum uORFs
uORF_Ge_START <- function(x){
  ifelse(as.character(strand(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]])[[1]])[1]=="+",start(unlist(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]]))[1],end(unlist(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]]))[1])
}
uORF_Ge_START_data <- sapply(seq_len(length(ORF_id)),function(x) uORF_Ge_START(x))

uORF_Ge_END <- function(x){
  ifelse(as.character(strand(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]])[[1]])[1]=="+",max(end(unlist(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]]))),min(start(unlist(fiveUTR_ORFs[Minimum_uORF$ORF_id[x]]))))
}
uORF_Ge_END_data <- sapply(seq_len(length(ORF_id)),function(x) uORF_Ge_END(x))
Minimum_uORF$uORF_Ge_range <- paste0(uORF_Ge_START_data,"-",uORF_Ge_END_data)

###Kallisto with 2 max isoforms
Kallisto_2MaxIso_out <- read.delim(file="~/Desktop/CTRL_TPC/Kallisto01272023/2MaxIsoforms/abundance.tsv",header=T,stringsAsFactors=F,sep="\t")
head(Kallisto_2MaxIso_out,30)
Kallisto_2MaxIso_out$tx_id <- Kallisto_2MaxIso_out$target_id
Kallisto_2MaxIso_out2 <- Kallisto_2MaxIso_out %>% dplyr::select(tx_id,tpm)

###
Minimum_uORF2 <- left_join(Minimum_uORF,Kallisto_2MaxIso_out2,by="tx_id")
Minimum_uORF2b <- Minimum_uORF2 %>% group_by(chr,uORF_Ge_range) %>% 
  mutate(tx_id_combine = toString(tx_id)) %>% 
  slice_max(order_by = tpm, n = 1) %>%  
  as.data.frame()
nrow(Minimum_uORF2b) #370 (10 in-frame and 50% in frame), 746 (1 in-frame) after remove redundant minimum uORFs on different isofroms


saveRDS(Minimum_uORF2b,"~/Desktop/CTRL_TPC/data/Rdata/Mar02_2023/2Max_isoform_0.5TPM_10reads_50perc_in_frame_1aa_ORFs.RData")
Minimum_uORF2c <- readRDS("~/Desktop/CTRL_TPC/data/Rdata/Mar02_2023/2Max_isoform_0.5TPM_10reads_50perc_in_frame_1aa_ORFs.RData")

Gene_description <- read.xlsx("~/Desktop/TAIR_gene_description_w_gene_id_Dec2020.xlsx",sheet=1)
uORF_ginfo <- dplyr::left_join(Minimum_uORF2c,Gene_description,by="gene_id")
uORF_ginfo$gene_model_type <- NULL
nrow(uORF_ginfo)
rownames(uORF_ginfo) <- NULL

uORF_ginfo <- uORF_ginfo %>% arrange(TotalC) %>% as.data.frame(stringsAsFactors=F)
colnames(uORF_ginfo)
uORF_ginfo$peptide_seq="M*"
uORF_ginfo$AA=1
uORF_ginfo2 <- uORF_ginfo %>% 
  dplyr::select("gene_id","ORF_id","tx_id_combine","Strand","uORF_Tx_range","uORF_Ge_range","InFrameC","TotalC","InFrameC_perc","InFrameS","TotalS","InFrameS_perc","AA", 
                "symbol","full_name",	"short_description",	"Curator_summary",	"Computational_description")
write.xlsx(uORF_ginfo,file = "~/Desktop/CTRL_TPC/Tables/Gene_description_370_minimum_uORF_2Max_isoform_0.5TPM_gt10reads_50perc_in_frame_Mar02_2023.xlsx")

```








